apple-silicon-support/packages/overlay.nix

```diff
@@ -0,0 +1,8 @@
+final: prev: {
+  linux-asahi = final.callPackage ./linux-asahi { };
+  m1n1 = final.callPackage ./m1n1 { };
+  uboot-asahi = final.callPackage ./uboot-asahi { };
+  asahi-fwextract = final.callPackage ./asahi-fwextract { };
+  mesa-asahi-edge = final.callPackage ./mesa-asahi-edge { };
```

Comment:
This needs to be `mesa-asahi-edge = final.callPackage ./mesa-asahi-edge { inherit (prev) mesa };` to prevent infinite recursion if the overlay installation method is used.

Reply:
I think setting `mesa = final.mesa-asahi-edge.override { inherit (prev) mesa; };` in the mesa module would be better.

Reply:
@ingenieroariel let me know if the latest commit has any issues

Reply:
Tried the latest commit and it still complains at:
```

       at /nix/store/pgyn69rm8wv2ppdpz04xxnr68003kphy-source/apple-silicon-support/packages/mesa-asahi-edge/default.nix:1:1:

            1| { lib
             | ^
            2| , fetchFromGitLab

       error: infinite recursion encountered

       at /nix/store/pgyn69rm8wv2ppdpz04xxnr68003kphy-source/apple-silicon-support/packages/mesa-asahi-edge/default.nix:5:2:

            4|
            5| (mesa.override {
             |  ^
             ```

Reply:
I think I'll go with depending on `prev.mesa` then.

(To be honest, I don't think overlays should be a nixos module option in general. It's too dependent on ordering for merging to make sense a lot of the time. I was mostly reluctant to use `prev` because of how little control users will have over the order in which overlays are applied. In my own configs, I just import nixpkgs from my flake, and `mkForce` overlays to `[]`.)

Reply:
As a user, that option would work for me. I guess the main thing is to make overlays possible and documented. Could you share an example flake.nix that does this?

The reason I need it is because firefox is taking 30 seconds on boot, and crashes all the time. I want to rule drivers issues out by compiling it from scratch.

Furthermore, once we get a flake.lock in the repo, we could share artifacts on cachix or a private server to use less electricity.

Reply:
The latest commit should work now (theoretically), and I'd appreciate it if you could confirm.

<details>
<summary>Here's a simplified version of my NixOS flake.nix. (click to expand)</summary>
<br>

```nix
{
  inputs = { stuff }; # nixpkgs, nixos-m1, etc

  outputs = inputs:
    let
      system = "aarch64-linux";

      pkgs = import inputs.nixpkgs {
        localSystem = { inherit system; };

        config = {
          allowUnfree = true;
        };

        overlays = [
          (final: prev: { }) # bring flake package outputs into pkgs

          inputs.nixos-m1.overlays.default
          inputs.another-flake.overlays.default

          (import ./packages/overlay.nix) # my own package overrides
        ];
      };
    in {
      nixosConfigurations.my-laptop = flakes.nixpkgs.lib.nixosSystem {
        inherit system pkgs;

        modules = [
          # disable module overlays
          { nixpkgs.overlays = inputs.nixpkgs.lib.mkForce []; }

          inputs.nixos-m1.nixosModules.default
          inputs.another-flake.nixosModules.default

          ./configuration.nix
        ];
      };
    };
}
```
</details>
<br>

> Furthermore, once we get a flake.lock in the repo, we could share artifacts on cachix or a private server to use less electricity.

We could use cachix to maybe eliminate the need for `hardware.asahi.pkgs`. I'm not sure how well that would work with non-flake systems, though.

Reply:
I confirm that the change to prev makes the overlay option work.

My computer crashes when compiling firefox but it is unrelated to this, sway and helix worked fine.